### DISCUSSION TITLE
Add change_scene_to_node() method

### DISCUSSION THREAD

**Discussion by sky94520** [godotengine/godot-proposals#9250] [Engine Core]
### Describe the project you are working on

A 2D survivor game at mobile platform using c#

### Describe the problem or limitation you are having in your project

Godot has change_scene_to_file and change_scene_to_packed methods, change_scene_to_packed. Whyt not add method change_scene_to_node(const Ref<Node> &p_node). 
Now I use sub thread load and instance scene, and add_child, set_current_scene. but the replacement is raised error at Android Platform which too many
```
Condition 'p_elem->_root != this' is true
```
I guess the godot source code has another logic while change scene.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I can use the new method to change_scene

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

No

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No

### Is there a reason why this should be core and not an add-on in the asset library?

No

**Comment by AThousandShips** [comment]
Are you adding the node to the scene on the thread as well? That's [not supported](https://docs.godotengine.org/en/stable/tutorials/performance/thread_safe_apis.html#scene-tree), you also can't load the same file in multiple threads

**Reply by sky94520** [reply]
Thank your reply. I don't change scene at thread. 
Firstly, I load and instance scene at thread, and call_deffered to change_scene. 
Because Godot doesn't method like change_scene_to_node, so I do use set current_scene like this
```
	var tree : SceneTree = self.get_tree()
	# Remove the old scene
	var old_scene = tree.current_scene
	
	tree.root.remove_child(old_scene)
	old_scene.queue_free()

	# Add the new scene
	tree.root.add_child(new_scene)
```
The above code run normally at first, and will raise error and the game will crashed at Android.
The Godot console raise many same error:
```
Condition 'p_elem->_root != this' is true
```

**Comment by Mickeon** [comment]
> ### If this enhancement will not be used often, can it be worked around with a few lines of script?
> No

To be pedantic, I believe yes?

```GDScript
get_tree().current_node.free()
get_tree().current_node = new_node
```

**Comment by dalexeev** [comment]
> To be pedantic, I believe yes?
> 
> ```gdscript
> get_tree().current_node.free()
> get_tree().current_node = new_node
> ```

> Condition 'p_elem->_root != this' is true

```gdscript
func change_scene_to_node(node: Node) -> void:
    await get_tree().process_frame
    get_tree().current_scene.free()
    get_tree().root.add_child(node)
    get_tree().current_scene = node
```

**Reply by sky94520** [reply]
Yes, you are right, The above code run normally at first, and will raise error and the game will crashed at Android.
The Godot console raise many same error:
```
Condition 'p_elem->_root != this' is true
```

**Reply by sky94520** [reply]
<img width="789" alt="image" src="https://github.com/godotengine/godot-proposals/assets/37724405/b257eb28-0fec-4b67-88bd-da1740efbec9">

**Reply by sky94520** [reply]
This error occurs after I switch scenes multiple times at Android.

**Reply by AThousandShips** [reply]
Does this happen on [4.3.dev4](https://godotengine.org/article/dev-snapshot-godot-4-3-dev-4/)? Are you using `TileMap`?

**Reply by sky94520** [reply]
I am sorry, My Godot Version is Godot 4.2.1 stable C#. 
Besides, you are right, I do use TileMap. The error is raised because TileMap? I want to use 4.3dev4, but now official doesn't Android C# export template.

**Reply by AThousandShips** [reply]
That's the issue then, the issue you're facing is a bug, solved by:
* https://github.com/godotengine/godot/pull/87478

It will be in a future 4.2 version as well, done in:
* https://github.com/godotengine/godot/pull/88916

**Reply by sky94520** [reply]
Looking forward to the arrival of the new version.
Thank you very much!!!

**Reply by AThousandShips** [reply]
If you experience any other problems after those fixes have been added to a release you use, please report it as a bug :)

**Comment by Calinou** [comment]
Moving to discussion, as there is no concrete technical implementation provided.
